src/scenic/core/regions.py

```diff
@@ -281,883 +314,2885 @@ def __hash__(self):
+
+        return PolygonalRegion(polygon=buffered_polygons, name=self.name, z=self.z)
+
+    @property
+    def dimensionality(self):
+        return 2
+
+    @cached_property
+    def size(self):
+        return self.polygons.area
+
+    @property
+    def area(self):
+        return self.size
+
+    def show(self, plt, style='r-', **kwargs):
+        plotPolygon(self.polygons, plt, style=style, **kwargs)
+
+    def __repr__(self):
+        return f'PolygonalRegion({self.polygons!r}, {self.z!r})'
+
+    def __eq__(self, other):
+        if type(other) is not PolygonalRegion:
+            return NotImplemented
+        return (other.polygons == self.polygons and
+                self.z == other.z and
+                other.orientation == self.orientation)
+
+    @cached
+    def __hash__(self):
+        return hash((self.polygons, self.orientation, self.z))
+
+class CircularRegion(PolygonalRegion):
+    """A circular region with a possibly-random center and radius.
+
+    Args:
+        center (`Vector`): center of the disc.
+        radius (float): radius of the disc.
+        resolution (int; optional): number of vertices to use when approximating this region as a
+            polygon.
+        name (str; optional): name for debugging.
+    """
+    def __init__(self, center, radius, resolution=32, name=None):
+        self.center = toVector(center, "center of CircularRegion not a vector")
+        self.radius = toScalar(radius, "radius of CircularRegion not a scalar")
+        self.resolution = resolution
+        self.circumcircle = (self.center, self.radius)
+
+        deps = [self.center, self.radius]
+
+        super().__init__(polygon=self._makePolygons(center, radius, resolution),
+            z=self.center.z, name=name, additionalDeps=deps)
+
+    @staticmethod
+    @distributionFunction
+    def _makePolygons(center, radius, resolution):
+        ctr = shapely.geometry.Point(center)
+        return ctr.buffer(radius, resolution=resolution)
+
+    ## Lazy Construction Methods ##
+    def sampleGiven(self, value):
+        return CircularRegion(value[self.center], value[self.radius],
+                              name=self.name, resolution=self.resolution)
+
+    def evaluateInner(self, context):
+        center = valueInContext(self.center, context)
+        radius = valueInContext(self.radius, context)
+        return CircularRegion(center, radius, name=self.name, resolution=self.resolution)
+
+    def intersects(self, other, triedReversed=False):
+        if isinstance(other, CircularRegion):
+            return self.center.distanceTo(other.center) <= self.radius + other.radius
+        
+        return super().intersects(other, triedReversed)
+
+    def containsPoint(self, point):
+        point = toVector(point)
+
+        if point.z != self.z:
+            return False
+
+        return point.distanceTo(self.center) <= self.radius
+
+    def distanceTo(self, point):
+        point = toVector(point)
+
+        if point.z == 0:
```

Comment:
Should compare to `self.z`. But rather than falling back to `super()` when the point isn't in the same plane, why not compute the distance exactly? I think you can do something like:
```python
dist2D = max(0, self.center.distanceTo(Vector(point.x, point.y, self.z)) - self.radius)
return math.hypot(dist2D, self.z - point.z)
```